### PR TITLE
fix for launching kernel invalid config error when calling embedding …

### LINKED ISSUE #130806: [rocm] F.embedding reports invalid configuration argument
### 🐛 Describe the bug

F.embedding will crash with relatively large tensor input on the AMD GPU: 

```
input = torch.randint(low=0, high=16032, size=[131072], device="cuda")
w = torch.randn([16032, 16384], device="cuda")
torch.nn.functional.embedding(input, w)

RuntimeError: HIP error: invalid configuration argument

```

### Versions

top of tree

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### PR DESCRIPTION
**Pull request by hongxiayang**
…with large index

Fixes #130806
When an output size of 2147483648 (=131072*16384) is expected in the above issue, it throwed out the following error:  
RuntimeError: HIP error: invalid configuration argument

What happened was that the second parameter passed to hipLaunchKernel was crazy {2147483648,1,1}.
Found issues in the Indexing.cu: 
 On ROCm, std::min -> ::min did not work as expected when outTotalSize>=2147483648

As the result, 2147483648 was sent to hipLaunchKernel which the GPU does not support such a huge number since this number specifies the number of threads per block. The original code intended to set 128 threads per block, though this is debatable as the perf would not good for latest powerful GPUs (a TODO item to update for perf maybe?) , but at least it would not cause `invalid configuration argument` error.

[Test] 
Run the same code snippet in the [issue](https://github.com/pytorch/pytorch/issues/130806), and print the output, its dim and numel(), which looks like below now:
```
output=tensor([[ 0.4044, -0.0244, -0.6865,  ..., -0.7800,  0.1175,  1.6726],
        [-1.0866, -0.1609,  0.3538,  ...,  1.9105,  0.7882,  1.1583],
        [-2.2079,  0.3736,  0.3610,  ..., -0.2658, -0.0459,  1.3077],
        ...,
        [ 0.8753, -0.7482, -0.1978,  ...,  0.9016,  1.1501, -0.5178],
        [-1.5845, -0.6277,  1.4520,  ...,  0.5733, -2.1198, -0.0915],
        [-0.6310, -1.0239, -0.1910,  ...,  0.4309,  0.1630,  0.3239]],
       device='cuda:0'), dim=2, numel=2147483648
```

Added a large tensor unit test too.
```
/pytorch# pytest test/nn/test_embedding.py -k test_large_tensors
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.9.19, pytest-7.3.2, pluggy-1.4.0
rootdir: /dockerx/development/pytorch
configfile: pytest.ini
plugins: flakefinder-1.1.0, rerunfailures-14.0, xdist-3.3.1, xdoctest-1.1.0, cpp-2.3.0, hypothesis-5.35.1
collected 288 items / 287 deselected / 1 selected                                                                                                                                        
Running 1 items in this shard

test/nn/test_embedding.py .                                                                                                                                                        [100%]

=========================================================================== 1 passed, 287 deselected in 3.16s ============================================================================
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 1]
```diff
@@ -900,10 +900,10 @@ void index_add_cuda_impl(const Tensor& self, int64_t dim, const Tensor& index, c
   // total size of the tensor ignoring dimension `dim`;
   // -the number of index we are choosing, which is the total size
   // of the tensor `index`.
-  const ptrdiff_t sliceSize = getSliceSize(self_, dim, index, source_);
-  const ptrdiff_t sourceTotalSize = source.numel();
+  const size_t sliceSize = getSliceSize(self_, dim, index, source_);
+  const long sourceTotalSize = source.numel();
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/Indexing.cu:904]: Avoid long, for consistency?  Other calls to numel store into int64_t.
```suggestion
  const int64_t sourceTotalSize = source.numel();
```

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 2]
```diff
@@ -900,10 +900,10 @@ void index_add_cuda_impl(const Tensor& self, int64_t dim, const Tensor& index, c
...
   const int64_t selfAddDimSize = self_.size(dim);
-  const ptrdiff_t numIndex = index.numel();
+  const long numIndex = index.numel();
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/Indexing.cu:906]:
```suggestion
  const int64_t numIndex = index.numel();
```

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 3]
```diff
@@ -1342,7 +1342,7 @@ void index_select_out_cuda_impl(
     const Tensor& self,
     long dim,
     const Tensor& index) {
-  ptrdiff_t numIndices = index.numel();
+  long numIndices = index.numel();
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/Indexing.cu:1345]: int64_t?

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 4]
```diff
@@ -1364,7 +1364,7 @@ void index_select_out_cuda_impl(
     at::native::resize_output(out, newSize);
   }
 
-  ptrdiff_t outTotalSize = out.numel();
+  long outTotalSize = out.numel();
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/Indexing.cu:1367]: int64_t?

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 5]
```diff
@@ -1377,7 +1377,7 @@ void index_select_out_cuda_impl(
   // -the number of indices we are choosing, which is the total size
   // of the tensor `indices`.
   int64_t selfSelectDimSize = self.dim() == 0 ? 1 : self.size(dim);
-  ptrdiff_t sliceSize = outTotalSize / numIndices;
+  long sliceSize = outTotalSize / numIndices;
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/Indexing.cu:1380]: int64_t?

**Review by jeffdaily — CHANGES_REQUESTED** [review]
Admittedly, I don't yet understand what went wrong and why this fixes it.  But I do see some use of long and some of int64_t, so perhaps settle on using int64_t?

File: aten/src/ATen/native/cuda/Indexing.cu (hunk 1 above)
Comment (hongxiayang) [reply aten/src/ATen/native/cuda/Indexing.cu:904]: > Avoid long, for consistency? Other calls to numel store into int64_t.
This file has many types, ptrdiff_t (not sure why use this point diff type to store sizes of tensors or number of elements of tensors), long (The signature of `numel` returns `long` atm), and int64_t. 
Use `int64_t` is ok and unsigned 64 bit integer `uint64_t` is even better as I don't think we expect the number of elements or size of tensors to be negative integers.

**Comment by hongxiayang** [comment]
> Admittedly, I don't yet understand what went wrong and why this fixes it. But I do see some use of long and some of int64_t, so perhaps settle on using int64_t?

I updated the description in the PR about what went wrong, hoping that helps for the understanding.

**Comment by hongxiayang** [comment]
There is more opportunity to refactor the code to make it better. Will leave it for future work as this is an urgent ask.

**Review by eqy — COMMENTED** [review]
Is it possible to add a test (perhaps decorated with @largeTensorTest) if necessary for this case

**Comment by hongxiayang** [comment]
Added.

**Review by jeffdaily — APPROVED** [review]
Approve if CI signal is good.

**Comment by hongxiayang** [comment]
> Approve if CI signal is good.

The only failing one is that Meta internal Diff is not in sync with external PR as I added a unit test as suggested by @eqy  after @xw285cornell imported it. @xw285cornell Please import again to make it sync. Thanks.

**Review by xw285cornell — APPROVED** [review]

**Comment by xw285cornell** [comment]
@hongxiayang @jeffdaily thanks for the fix! 

Just wondering, "2: On ROCm, std::min -> ::min did not work as expected when outTotalSize>=2147483648", shall we fix this in rocm?

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 6]
```diff
@@ -1073,11 +1073,11 @@ void index_reduce_func_cuda_impl(
   // total size of the tensor ignoring dimension `dim`;
   // -the number of index we are choosing, which is the total size
   // of the tensor `index`.
-  ptrdiff_t sliceSize = getSliceSize(self_, dim, index, source_);
-  ptrdiff_t sourceTotalSize = source.numel();
-  int64_t selfReduceDimSize = self_.size(dim);
-  ptrdiff_t numIndex = index.numel();
-  int64_t selfNumel = self_.numel();
+  uint64_t sliceSize = getSliceSize(self_, dim, index, source_);
```
Comment (xintwfb) [inline aten/src/ATen/native/cuda/Indexing.cu:1076]: Thanks @hongxiayang for the fix! 
In the description you mentioned 1 and 2, and the diff looks mainly fix 1 by using uint64_t which has larger representable range. But I just wondering if we actually should fix `2: On ROCm, std::min -> ::min did not work as expected` instead? (although uint64_t has much larger representable, but in a extreme, in the future if our LLM context length continue to grow, will it to a point that beyond uint64_t representable range?)

**Comment by hongxiayang** [comment]
> @hongxiayang @jeffdaily thanks for the fix!
> 
> Just wondering, "2: On ROCm, std::min -> ::min did not work as expected when outTotalSize>=2147483648", shall we fix this in rocm?

will follow up on this. thanks.

**Pull request closed without merging**